In [42]:
# Import libraries

import numpy as np
import pandas as pd

In [43]:
# Import datasets

dataset = pd.read_csv('explored_data.csv')
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, -1].values

In [44]:
print(X)

[[3 0 0 ... 'Private' 'Registered' 'All Day of The Week']
 [0 0 0 ... 'Private' 'Unknown' 'All Day of The Week']
 [1 0 0 ... 'Private' 'Registered' 'All Day of The Week']
 ...
 [0 0 0 ... 'Public' 'Unknown'
  'All Day of The Week Except Sunday, Saturday']
 [0 0 0 ... 'Public' 'Unknown' 'All Day of The Week']
 [0 0 0 ... 'Public' 'Unknown'
  'All Day of The Week Except Sunday, Saturday']]


In [45]:
print(y)

[1 0 0 ... 0 0 0]


In [46]:
# Convert the NumPy array to a pandas DataFrame

df = pd.DataFrame(X)

In [47]:
df

0  1  2  3  4  5  6    7        8        9           10  \
0      3  0  0  2  1  1  2   14     Abia  Private  Registered   
1      0  0  0  0  0  0  0    5     Abia  Private     Unknown   
2      1  0  0  1  0  1  1   17     Abia  Private  Registered   
3      2  0  0  1  0  1  0   12     Abia  Private  Registered   
4      0  0  0  0  0  0  0    5     Abia  Private     Unknown   
...   .. .. .. .. .. .. ..  ...      ...      ...         ...   
36705  2  1  2  4  1  3  1  104  Zamfara   Public     Unknown   
36706  0  0  0  0  0  0  0    5  Zamfara   Public     Unknown   
36707  0  0  0  0  0  0  0    6  Zamfara   Public     Unknown   
36708  0  0  0  0  0  2  0   17  Zamfara   Public     Unknown   
36709  0  0  0  0  0  0  0    5  Zamfara   Public     Unknown   

                                                11  
0                              All Day of The Week  
1                              All Day of The Week  
2                              All Day of The Week  
3                              All Day of The Week  
4                              All Day of The Week  
...                                            ...  
36705                          All Day of The Week  
36706  All Day of The Week Except Sunday, Saturday  
36707  All Day of The Week Except Sunday, Saturday  
36708                          All Day of The Week  
36709  All Day of The Week Except Sunday, Saturday  

[36710 rows x 12 columns]

In [48]:
column_to_convert = [8, 9, 10, 11]
df = pd.get_dummies(df, columns=column_to_convert)

In [49]:
df

0  1  2  3  4  5  6    7  8_Abia  8_Adamawa  ...  11_On Wednesday  \
0      3  0  0  2  1  1  2   14    True      False  ...            False   
1      0  0  0  0  0  0  0    5    True      False  ...            False   
2      1  0  0  1  0  1  1   17    True      False  ...            False   
3      2  0  0  1  0  1  0   12    True      False  ...            False   
4      0  0  0  0  0  0  0    5    True      False  ...            False   
...   .. .. .. .. .. .. ..  ...     ...        ...  ...              ...   
36705  2  1  2  4  1  3  1  104   False      False  ...            False   
36706  0  0  0  0  0  0  0    5   False      False  ...            False   
36707  0  0  0  0  0  0  0    6   False      False  ...            False   
36708  0  0  0  0  0  2  0   17   False      False  ...            False   
36709  0  0  0  0  0  0  0    5   False      False  ...            False   

       11_Thursday & Friday  11_Thursday & Monday  11_Thursday & Sunday  \
0                     False                 False                 False   
1                     False                 False                 False   
2                     False                 False                 False   
3                     False                 False                 False   
4                     False                 False                 False   
...                     ...                   ...                   ...   
36705                 False                 False                 False   
36706                 False                 False                 False   
36707                 False                 False                 False   
36708                 False                 False                 False   
36709                 False                 False                 False   

       11_Thursday & Tuesday  11_Thursday & Wednesday  11_Tuesday & Friday  \
0                      False                    False                False   
1                      False                    False                False   
2                      False                    False                False   
3                      False                    False                False   
4                      False                    False                False   
...                      ...                      ...                  ...   
36705                  False                    False                False   
36706                  False                    False                False   
36707                  False                    False                False   
36708                  False                    False                False   
36709                  False                    False                False   

       11_Tuesday & Monday  11_Tuesday & Saturday  11_Tuesday & Wednesday  
0                    False                  False                   False  
1                    False                  False                   False  
2                    False                  False                   False  
3                    False                  False                   False  
4                    False                  False                   False  
...                    ...                    ...                     ...  
36705                False                  False                   False  
36706                False                  False                   False  
36707                False                  False                   False  
36708                False                  False                   False  
36709                False                  False                   False  

[36710 rows x 108 columns]

In [50]:
X = df.to_numpy()

In [51]:
print(X)

[[3 0 0 ... False False False]
 [0 0 0 ... False False False]
 [1 0 0 ... False False False]
 ...
 [0 0 0 ... False False False]
 [0 0 0 ... False False False]
 [0 0 0 ... False False False]]


In [52]:
# Split the dataset to train and test sets

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 1)

In [53]:
print(X_train)

[[0 0 0 ... False False False]
 [0 0 0 ... False False False]
 [0 0 0 ... False False False]
 ...
 [0 0 0 ... False False False]
 [0 0 0 ... False False False]
 [0 0 0 ... False False False]]


In [54]:
print(X_test)

[[1 0 0 ... False False False]
 [0 0 0 ... False False False]
 [0 0 0 ... False False False]
 ...
 [0 0 0 ... False False False]
 [0 0 0 ... False False False]
 [1 0 1 ... False False False]]


In [55]:
print(y_train)

[0 0 0 ... 0 0 0]


In [56]:
print(y_test)

[0 0 0 ... 1 0 0]


In [57]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train[:, :7] = sc.fit_transform(X_train[:, :7])
X_test[:, :7] = sc.transform(X_test[:, :7])

In [58]:
print(X_train)

[[-0.07417056730471862 -0.0845550949220358 -0.032747605396165516 ...
  False False False]
 [-0.07417056730471862 -0.0845550949220358 -0.032747605396165516 ...
  False False False]
 [-0.07417056730471862 -0.0845550949220358 -0.032747605396165516 ...
  False False False]
 ...
 [-0.07417056730471862 -0.0845550949220358 -0.032747605396165516 ...
  False False False]
 [-0.07417056730471862 -0.0845550949220358 -0.032747605396165516 ...
  False False False]
 [-0.07417056730471862 -0.0845550949220358 -0.032747605396165516 ...
  False False False]]


In [59]:
print(X_test)

[[0.030321480119486325 -0.0845550949220358 -0.032747605396165516 ...
  False False False]
 [-0.07417056730471862 -0.0845550949220358 -0.032747605396165516 ...
  False False False]
 [-0.07417056730471862 -0.0845550949220358 -0.032747605396165516 ...
  False False False]
 ...
 [-0.07417056730471862 -0.0845550949220358 -0.032747605396165516 ...
  False False False]
 [-0.07417056730471862 -0.0845550949220358 -0.032747605396165516 ...
  False False False]
 [0.030321480119486325 -0.0845550949220358 0.13404595604583738 ... False
  False False]]


# Logistic Regression

In [60]:
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state = 0, max_iter=1000)
classifier.fit(X_train, y_train)

LogisticRegression(max_iter=1000, random_state=0)

In [61]:
y_pred = classifier.predict(X_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

[[0 0]
 [0 0]
 [0 0]
 ...
 [0 1]
 [0 0]
 [0 0]]


In [62]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[7029   31]
 [ 233   49]]


0.9640424952329065

In [70]:
def calculate_metrics(cm):
    # Extract TP, TN, FP, FN from the confusion matrix
    TP, FP = cm[0]
    FN, TN = cm[1]

    # Calculate precision, recall, and F1 score
    precision = TP / (TP + FP)
    recall = TP / (TP + FN)
    f1_score = 2 * (precision * recall) / (precision + recall)

    # Print the metrics
    print(f"Precision: {precision}")
    print(f"Recall: {recall}")
    print(f"F1 Score: {f1_score}")

In [71]:
calculate_metrics(cm)

Metrics for Model 1:
Precision: 0.9956090651558074
Recall: 0.9679151748829523
F1 Score: 0.9815668202764978


# K-Nearest Neighbors (K-NN)

In [72]:
from sklearn.neighbors import KNeighborsClassifier
classifier_2 = KNeighborsClassifier(n_neighbors = 5, metric = 'minkowski', p = 2)
classifier_2.fit(X_train, y_train)

KNeighborsClassifier()

In [73]:
y_pred_2 = classifier_2.predict(X_test)
print(np.concatenate((y_pred_2.reshape(len(y_pred_2),1), y_test.reshape(len(y_test),1)),1))

[[0 0]
 [0 0]
 [0 0]
 ...
 [0 1]
 [0 0]
 [0 0]]


In [74]:
cm = confusion_matrix(y_test, y_pred_2)
print(cm)
accuracy_score(y_test, y_pred_2)

[[6984   76]
 [ 200   82]]


0.9624080631980387

In [75]:
def calculate_metrics(cm):
    # Extract TP, TN, FP, FN from the confusion matrix
    TP, FP = cm[0]
    FN, TN = cm[1]

    # Calculate precision, recall, and F1 score
    precision = TP / (TP + FP)
    recall = TP / (TP + FN)
    f1_score = 2 * (precision * recall) / (precision + recall)

    # Print the metrics
    print(f"Precision: {precision}")
    print(f"Recall: {recall}")
    print(f"F1 Score: {f1_score}")

In [76]:
calculate_metrics(cm)

Metrics for Model 1:
Precision: 0.9892351274787535
Recall: 0.9721603563474388
F1 Score: 0.9806234203875316


# Kernel SVM

In [77]:
from sklearn.svm import SVC
classifier_3 = SVC(kernel = 'rbf', random_state = 0)
classifier_3.fit(X_train, y_train)

SVC(random_state=0)

In [78]:
y_pred_3 = classifier_3.predict(X_test)
print(np.concatenate((y_pred_3.reshape(len(y_pred_3),1), y_test.reshape(len(y_test),1)),1))

[[0 0]
 [0 0]
 [0 0]
 ...
 [0 1]
 [0 0]
 [0 0]]


In [79]:
cm = confusion_matrix(y_test, y_pred_3)
print(cm)
accuracy_score(y_test, y_pred_3)

[[7027   33]
 [ 237   45]]


0.9632252792154726

In [80]:
def calculate_metrics(cm):
    # Extract TP, TN, FP, FN from the confusion matrix
    TP, FP = cm[0]
    FN, TN = cm[1]

    # Calculate precision, recall, and F1 score
    precision = TP / (TP + FP)
    recall = TP / (TP + FN)
    f1_score = 2 * (precision * recall) / (precision + recall)

    # Print the metrics
    print(f"Precision: {precision}")
    print(f"Recall: {recall}")
    print(f"F1 Score: {f1_score}")

In [81]:
calculate_metrics(cm)

Metrics for Model 1:
Precision: 0.9953257790368272
Recall: 0.9673733480176211
F1 Score: 0.9811505166154705


# Naive Bayes

In [82]:
from sklearn.naive_bayes import GaussianNB
classifier_4 = GaussianNB()
classifier_4.fit(X_train, y_train)

GaussianNB()

In [83]:
y_pred_4 = classifier_4.predict(X_test)
print(np.concatenate((y_pred_4.reshape(len(y_pred_4),1), y_test.reshape(len(y_test),1)),1))

[[1 0]
 [1 0]
 [1 0]
 ...
 [1 1]
 [1 0]
 [1 0]]


In [84]:
cm = confusion_matrix(y_test, y_pred_4)
print(cm)
accuracy_score(y_test, y_pred_4)

[[ 201 6859]
 [   0  282]]


0.0657858894034323

In [85]:
def calculate_metrics(cm):
    # Extract TP, TN, FP, FN from the confusion matrix
    TP, FP = cm[0]
    FN, TN = cm[1]

    # Calculate precision, recall, and F1 score
    precision = TP / (TP + FP)
    recall = TP / (TP + FN)
    f1_score = 2 * (precision * recall) / (precision + recall)

    # Print the metrics
    print(f"Precision: {precision}")
    print(f"Recall: {recall}")
    print(f"F1 Score: {f1_score}")

In [86]:
calculate_metrics(cm)

Metrics for Model 1:
Precision: 0.02847025495750708
Recall: 1.0
F1 Score: 0.05536427489326539


# Decision Tree Regression

In [87]:
from sklearn.tree import DecisionTreeClassifier
classifier_5 = DecisionTreeClassifier(criterion = 'entropy', random_state = 0)
classifier_5.fit(X_train, y_train)

DecisionTreeClassifier(criterion='entropy', random_state=0)

In [88]:
y_pred_5 = classifier_5.predict(X_test)
print(np.concatenate((y_pred_5.reshape(len(y_pred_5),1), y_test.reshape(len(y_test),1)),1))

[[0 0]
 [0 0]
 [0 0]
 ...
 [0 1]
 [0 0]
 [0 0]]


In [89]:
cm = confusion_matrix(y_test, y_pred_6)
print(cm)
accuracy_score(y_test, y_pred_6)

[[6896  164]
 [ 179  103]]


0.953282484336693

In [90]:
def calculate_metrics(cm):
    # Extract TP, TN, FP, FN from the confusion matrix
    TP, FP = cm[0]
    FN, TN = cm[1]

    # Calculate precision, recall, and F1 score
    precision = TP / (TP + FP)
    recall = TP / (TP + FN)
    f1_score = 2 * (precision * recall) / (precision + recall)

    # Print the metrics
    print(f"Precision: {precision}")
    print(f"Recall: {recall}")
    print(f"F1 Score: {f1_score}")

In [91]:
calculate_metrics(cm)

Precision: 0.976770538243626
Recall: 0.9746996466431095
F1 Score: 0.9757339936328263


# Random Forest Classification

In [92]:
from sklearn.ensemble import RandomForestClassifier
classifier_6 = RandomForestClassifier(n_estimators = 10, criterion = 'entropy', random_state = 0)
classifier_6.fit(X_train, y_train)

RandomForestClassifier(criterion='entropy', n_estimators=10, random_state=0)

In [93]:
y_pred_6 = classifier_6.predict(X_test)
print(np.concatenate((y_pred_6.reshape(len(y_pred_6),1), y_test.reshape(len(y_test),1)),1))

[[0 0]
 [0 0]
 [0 0]
 ...
 [0 1]
 [0 0]
 [0 0]]


In [94]:
cm = confusion_matrix(y_test, y_pred_6)
print(cm)
accuracy_score(y_test, y_pred_6)

[[6992   68]
 [ 217   65]]


0.9611822391718877

In [95]:
def calculate_metrics(cm):
    # Extract TP, TN, FP, FN from the confusion matrix
    TP, FP = cm[0]
    FN, TN = cm[1]

    # Calculate precision, recall, and F1 score
    precision = TP / (TP + FP)
    recall = TP / (TP + FN)
    f1_score = 2 * (precision * recall) / (precision + recall)

    # Print the metrics
    print(f"Precision: {precision}")
    print(f"Recall: {recall}")
    print(f"F1 Score: {f1_score}")

In [96]:
calculate_metrics(cm)

Precision: 0.9903682719546743
Recall: 0.9698987376889998
F1 Score: 0.9800266311584555


I will have to choose Logistic Regression because it has the an accuracy score of 0.9640424952329065 which is 96.4%, but that is not the main reason because there are other three models with the same 90+%. The main reason i choose Logistic Regression is because it has the highest f1 score, and why i am using the f1 score instead of the precision and recall is because i want to make sure the model have a balance between precision and recall. Because we do not want to falsely predict a facility that provides ambulance services when it doesn't (precision), and also we do not want to miss a facility that provides ambulance services (recall) as well.